casper/src/main/scala/coop/rchain/casper/engine/Running.scala

```diff
@@ -281,28 +281,111 @@ class Running[F[_]: Sync: BlockStore: CommUtil: TransportLayer: ConnectionsCell:
   private val F    = Applicative[F]
   private val noop = F.unit
 
+  /**
+    * Message that relates to a block (e.g. block message or block hash broadcast message) shall be considered
+    * as repeated and dropped if corresponding block is marked as received in RequestedBlocks or already added to DAG.
+    * Otherwise - that is a new message and shall be processed normal way.
+    *
+    * RequestedBlocks is a part of Casper loop maintaining liveness of the protocol.
+    * It keeping records of blocks being requested and re-request them on schedule if failed first time.
+    * It is expected to get lots of duplicate messages from different peers, so check for duplicate is required.
+    * Processing of duplicate blocks puts load on Casper and highly discouraged.
+    *
+    * When block is received from some peer first time - it is
+    * - marked as `received` in RequestedBlocks,
+    * - added to the block store and
+    * - sent to Casper.
+    * Block is being kept in RequestedBlocks until it is added to the DAG so we can detect duplicates.
+    *
+    * @param hash Block hash
+    * @return If message is repeated
+    */
+  private def repeatedCasperMessage(hash: BlockHash): F[Boolean] =
+    // If either RequestedBlocks shows that block was received (but not yet added to Casper)
+    // or block was already added by Casper (so was removed from RequestedBlocks)
+    (RequestedBlocks.get(hash).map(x => x.nonEmpty && x.get.received) ||^ casper.contains(hash))
+      .ifM(
+        // message is repeated
+        true.pure[F],
+        false.pure[F]
+      )
```

Comment:
No need for `ifM`, `_ ||^ _: F[Boolean]`.

Reply:
https://github.com/rchain/rchain/pull/2902/commits/372522ce494a5493354b7a3505316204e77eb715